KT-16661

qwwdfsad opened the pull request:
Performance improvements of `Strings#split`.
For detailed evaluation see [KT-22248](https://youtrack.jetbrains.com/issue/KT-22248) (which is now marked as duplicate of KT-16661)

After 25c6cb0 Kotlin version became consistently faster than Java version for any arguments.

```
Benchmark                 (needle)   Mode  Cnt   Score   Error   Units
SplitBenchmark.splitJava         ,  thrpt   10  57.818 ± 1.740  ops/us
SplitBenchmark.splitJava         a  thrpt   10   6.687 ± 1.353  ops/us
SplitBenchmark.splitKt           ,  thrpt   10  77.459 ± 1.855  ops/us
SplitBenchmark.splitKt           a  thrpt   10  11.378 ± 0.299  ops/us

```

ilya-g commented on libraries/stdlib/src/kotlin/text/Strings.kt:
```diff
@@ -1183,8 +1191,53 @@ public fun CharSequence.splitToSequence(vararg delimiters: Char, ignoreCase: Boo
+    var currentOffset = 0
+    var nextIndex = indexOf(delimiter, currentOffset, ignoreCase)
+    val isLimited = limit > 0
+    val result = ArrayList<String>()
```
Agree, please do, but only for `isLimited` case.
Better to use `minOf`, it's available without import.

qwwdfsad replied:
Why not in both cases?
10 (not 16 actually) is default array list size in both openjdk and android.

`result = ArrayList<String>(minOf(limit, 10))` looks simpler than `result = if (isLimited) ... else ...`

ilya-g replied:
`minOf(limit, 10)` is 0 when the limit is 0. Is the capacity intended to be so?
Actually it would be better to use `limit.coerceAtMost(10)` here to communicate the intent more clearly.

qwwdfsad replied:
Yes. "0" indicates that default size should be used

ilya-g replied:
As far as I can see in the sources, `ArrayList()` and `ArrayList(0)` constructors initialize the list state differently, the latter grows its capacity more often, so it may result in more reallocations.

qwwdfsad replied:
Yes, you're right

ilya-g commented on libraries/stdlib/src/kotlin/text/Strings.kt:
```diff
@@ -1163,8 +1163,16 @@ public fun CharSequence.splitToSequence(vararg delimiters: String, ignoreCase: B
+    if (delimiters.size == 1) {
+        val delimiter = delimiters[0]
+        if (!delimiter.isEmpty()) {
+            return split(delimiters[0], ignoreCase, limit)
```
`delimiters[0]` value is already in `delimiter` local val.

ilya-g commented on libraries/stdlib/src/kotlin/text/Strings.kt:
```diff
@@ -1183,8 +1191,56 @@ public fun CharSequence.splitToSequence(vararg delimiters: Char, ignoreCase: Boo
+            currentOffset = nextIndex + delimiter.length
+            nextIndex = indexOf(delimiter, currentOffset, ignoreCase)
+        } else {
+            result.add(substring(currentOffset, length))
```
Looks like we can return result right after this.

ilya-g replied:
Also if we get here, it means that `nextIndex` search at the previous iteration is unused.

qwwdfsad replied:
Simplified it a bit more, though `nextIndex` is still unused. It doesn't make any difference (in my small-medium string size benchmarks, though I'm sure it's still possible to show the opposite), but makes code less readable

ilya-g replied:
Consider the case when you want to split strings like `key=very_long_value` with limit=2. In this case the function will search for a second `=` to the end of a long string, though it could have stopped already.

qwwdfsad replied:
Point taken, fixed (and simplified again)

ilya-g commented on libraries/stdlib/src/kotlin/text/Strings.kt:
```diff
@@ -1183,8 +1191,50 @@ public fun CharSequence.splitToSequence(vararg delimiters: Char, ignoreCase: Boo
+
+    var currentOffset = 0
+    var nextIndex = indexOf(delimiter, currentOffset, ignoreCase)
+    if (nextIndex == -1) {
```
Looks like we can add ` || limit == 1` because it should return the same single item list.

ilya-g replied:
Though there isn't a lot of use cases for limit=1, thus we can avoid introducing a special path for this case.

ilya-g commented on libraries/stdlib/src/kotlin/text/Strings.kt:
```diff
@@ -1222,10 +1222,8 @@ private fun CharSequence.split(delimiter: String, ignoreCase: Boolean, limit: In
-            result.add(substring(currentOffset, length))
-            return result
+            // Do not search for next occurrence if we're reaching limit
+            nextIndex = if (result.size == limit - 1) -1 else indexOf(delimiter, currentOffset, ignoreCase)
```
`break` instead of `-1` would be more clear

ilya-g commented:
All tests have passed, so I'm going to merge this soon. Would you mind if I squash some commits?

Could you also share the code of benchmarks? A comment here or a gist would be enough.

qwwdfsad commented:
>Would you mind if I squash some commits?

Sure, I don't mind

>Could you also share the code of benchmarks?

I was playing with parameters in this one
```
@Fork(value = 1)//, jvmArgsAppend = arrayOf("-XX:+UnlockDiagnosticVMOptions", "-XX:+DebugNonSafepoints"))  //, jvmArgsAppend = arrayOf("-XX:TieredStopAtLevel=1"))
@Measurement(iterations = 10)
@Warmup(iterations = 10)
@State(Scope.Benchmark)
@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
open class SplitBenchmark {

    @Param("haystack")
    lateinit private var haystack: String

    @Param("a", ",")
    lateinit private var needle: String

    @Benchmark
    fun splitJava(): List<String> {
        return (haystack as java.lang.String).split(needle).toList()
    }

    @Benchmark
    fun splitKt(): List<String> {
        return haystack.split(needle)
    }

    @Benchmark
    fun splitKtLimit(): List<String> {
        return haystack.split(needle, limit = 2)
    }
}

```

ilya-g commented:
I've simplified it a bit more and merged, see the commits https://github.com/JetBrains/kotlin/compare/2805371bdcfa05bd2e72f86077f3a54518fd89ae~1...6b9520ec73c8a6c0f83d4ec3afd2a295364ed54a

Thank you for the PR.
